accepted/2021/Globalization/GlobalizationSize.md

```diff
@@ -0,0 +1,81 @@
+```
+
+- **Mobile Platforms**:
+  - **Android**: Although Android OS comes with the ICU libraries which .NET can use, the OEMs can choose to not include such libraries in their device images.
+  - **iOS/tvOS**: don't come with an ICU library that .NET can use. .NET had to provide an ICU package to be used in such platforms.
+- **Web assembly**: The clients have to be a small size and need to run inside the browser which will not allow accessing any libraries outside the browser. That means WebAssembly clients need to include an ICU package for globalization support.
+- Users of the **Alpine Linux containers** which enable `Globalization Invariant Mode` ask for better globalization support without increasing the image size much.
+
+### Current Status
+
+In .NET 5.0/6.0 we have introduced the ICU app-local feature which allowed the applications to publish any ICU package as part of the app. To solve the size issue of mobile apps and WebAssembly clients, Marek's team has worked to create size-trimmed ICU packages targeting different mobile platforms and WebAssembly clients. Although this was helpful to move forward, it still takes a lot of effort and needs more to be done. It became clear trimming ICU4C is not easy. It is not as simple as just trimming the data but the ICU code size was big and needs trimming too. Trimming ICU code is not a simple task. The code that thought is isolated and trimmed, caused some problems later as it was used for other parts. Looks full code analysis needed to ensure whatever we trim would be safe and not needed for other parts.
+
+One of the suggested thoughts is to try to enhance the Globalization Invariant mode and try to make it providing better globalization support. We already tracking doing some of this work in .NET 6.0  in the [issue](https://github.com/dotnet/runtime/issues/43774). Looking at the scenarios we want to enhance/support, I am not convinced enhancing the globalization mode is the path we need to pursue. The reason is, to address all concerned scenarios, we need to offer a range of different trimmed data and functionality options. For example, we need to offer packages for one or a group of cultures. Something like European-only cultures, CJK cultures, Bidi Cultures, full list of cultures...etc. Also, we need to provide functionality options e.g. Cultural collation support, Normalization support, IDN support, casing support...etc. If we can do that, the users can have the freedom to select the level of the functionality they want and would be clear the size cost for every choice. The users will have full control over the functionality level and the size cost. To support that, would make sense we follow the same ICU app-local idea and not just try to add more functionality to the Invariant mode. It would be better too to keep the definition of Globalization Invariant Mode clear and not confusing. This mode is for the applications that don't need globalization support and not for anything else.
```

Comment:
How does one tell whether or not the app and its dependencies need the different functionality options, like normalization support, etc.? It is nice to give full control to users; but we also need to have a guidance how to take advantage of this support with confidence.

Reply:
> How does one tell whether or not the app and its dependencies need the different functionality options, like normalization support, etc.? It is nice to give full control to users; but we also need to have a guidance how to take advantage of this support with confidence.

That is make sense: 
- The easy one would be the collation. If the user does not care about linguistic operation and ordinal operation is good enough then can exclude collation all together. 
- Normalization would be needed in advanced scenarios like transformation engines (e.g. text to speech, advanced DB searches, semantic string comparisons...etc.). Also, normalization is important part for IDN too.
- IDN support, which need normalization too for advanced scenario (to fully support non-ASCII URLs). 

I'll add this info to the doc

Reply:
The problem are nuget packages. How does the user knows what the nuget package needs?

For example, Uri parsing needs FormC normalization and I believe that it is slightly broken with invariant mode today. How does one discovers situations like this without trial and error?

Reply:
Another simple case to answer: Does Maui need collation? If basic Maui does not need collation, are there Maui controls that do need collation?

Reply:
The only way to do that is documenting what could bebroken when excluding any functionality. Maybe in the NuGet spec files itself. 

The other idea would have only 2 types of packages. One with the full functionality and the other without. i.e. the second would be missing collation, normalization...etc. That would be closer to Invariant mode except the package will carry some locale data.

Unfortunately, I don't have solid data. so, we may delay support cutting functionality till we get concrete requests for that.

Reply:
> Another simple case to answer: Does Maui need collation? If basic Maui does not need collation, are there Maui controls that do need collation?

I believe the UI apps in general will need globalization support. Usually, apps have a user facing controls which has a listed sorted in the user language. formatting and displaying dates and numbers is common too there. In general, I think Maui would be the first-class platform need globalization.